soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp

```diff
@@ -0,0 +1,1046 @@
+    Color_RGBA8 areaIncompleteColor = CVar_GetRGBA("gCheckTrackerAreaMainIncompleteColor", Color_Main_Default);
+    Color_RGBA8 extraCompleteColor = CVar_GetRGBA("gCheckTrackerAreaExtraCompleteColor", Color_Area_Complete_Extra_Default);
+    Color_RGBA8 extraIncompleteColor = CVar_GetRGBA("gCheckTrackerAreaExtraIncompleteColor", Color_Area_Incomplete_Extra_Default);
+    Color_RGBA8 mainColor;
+    Color_RGBA8 extraColor;
+    std::string stemp;
+    s32 areaMask = 1;
+
+    // Logic for each check
+    for (auto& obj : checks)
+    {
+
+        //New Area to be drawn
+        if (obj.rcArea != lastArea)
+        {
+            //Last Area needs to be cleaned up
+            if (lastArea != RCAREA_INVALID && doDraw) {
+                ImGui::TreePop();
+                UIWidgets::PaddedSeparator();
+            }
+            lastArea = obj.rcArea;
+
+            //Decide if we should skip because of hiding rules
+            thisAreaFullyChecked = ((areasFullyChecked & areaMask) != 0);
+            if (!showHidden && (
+                hideComplete && thisAreaFullyChecked ||
+                hideIncomplete && !thisAreaFullyChecked 
+            )) {
+                doDraw = false;
+            }
+            else {
+                //Get the colour for the area
+                if (thisAreaFullyChecked) {
+                    mainColor = areaCompleteColor;
+                    extraColor = extraCompleteColor;
+                } else {
+                    mainColor = areaIncompleteColor;
+                    extraColor = extraIncompleteColor;
+                }
+
+                //Draw the area
+                collapseLogic = !thisAreaFullyChecked;
+                if (doingCollapseOrExpand) {
+                    if (optExpandAll) {
+                        collapseLogic = true;
+                    } else if (optCollapseAll) {
+                        collapseLogic = false;
+                    }
+                }
+                stemp = RandomizerCheckObjects::GetRCAreaName(obj.rcArea) + "##TreeNode";
+                ImGui::PushStyleColor(ImGuiCol_Text, Color_RGBA8_to_ImVec4(mainColor));
+                if (doingCollapseOrExpand)
+                    ImGui::SetNextItemOpen(collapseLogic, ImGuiCond_Always);
+                else
+                    ImGui::SetNextItemOpen(!thisAreaFullyChecked, ImGuiCond_Once);
+                doDraw = ImGui::TreeNode(stemp.c_str());
+                ImGui::PopStyleColor();
+                ImGui::SameLine();
+                ImGui::PushStyleColor(ImGuiCol_Text, Color_RGBA8_to_ImVec4(extraColor));
+
+                isThisAreaSpoiled = areasSpoiled & areaMask || CVar_GetS32("gCheckTrackerOptionMQSpoilers", 0);
+
+                if (isThisAreaSpoiled) {
+                    if (showVOrMQ && RandomizerCheckObjects::AreaIsDungeon(obj.rcArea)) {
+                        if (OTRGlobals::Instance->gRandomizer->masterQuestDungeons.contains(dungeonSceneLookupByArea[obj.rcArea]))
```

Comment:
since `dungeonSceneLookupByArea` is only being used here, it might be nice to refactor that into a method with a switch/case or something, it's pretty hard to read that array as-is with the first 20 entries just being `SCENE_ID_MAX`

Reply:
Refactored in commit 35b0a6033ab28067afd707ed2cb29a01e17d723c